compiler/sem/semtypes.nim

```diff
@@ -747,80 +749,97 @@ proc checkForOverlap(c: PContext, t: PNode, currentEx, branchIndex: int) =
 
   result = newNodeI(nkRange, a.info)
-  result.add(at)
-  result.add(bt)
-  if emptyRange(ac, bc):
-    localReport(c.config, b, reportSem rsemRangeIsEmpty)
-  elif t[0].typ.skipTypes(abstractInst).kind == tyString:
-    # XXX: ``nkError`` needs to be used here
-    localReport(c.config, b, reportSem rsemStringRangeNotAllowed)
-  else: covered = covered + getOrdValue(bc) + 1 - getOrdValue(ac)
-
-proc semCaseBranchRange(c: PContext, t, b: PNode,
+  result.add(a)
+  result.add(b)
+  if a.kind == nkError or b.kind == nkError:
+    result = c.config.wrapError(result)
+  elif typ.skipTypes(abstractInst).kind == tyString:
+    result = c.config.newError(result):
+      PAstDiag(kind: adSemStringRangeNotAllowed)
+  elif emptyRange(a, b):
+    result = c.config.newError(result):
+      PAstDiag(kind: adSemRangeIsEmpty)
+  else:
+    # all good; no error
+    covered = covered + getOrdValue(b) + 1 - getOrdValue(a)
+
+proc semCaseBranchRange(c: PContext, typ: PType, b: PNode,
                         covered: var Int128): PNode =
   checkSonsLen(b, 3, c.config)
-  result = semBranchRange(c, t, b[1], b[2], covered)
-
-proc semCaseBranchSetElem(c: PContext, t, b: PNode,
-                          covered: var Int128): PNode =
-  if isRange(b):
-    checkSonsLen(b, 3, c.config)
-    result = semBranchRange(c, t, b[1], b[2], covered)
-  elif b.kind == nkRange:
-    checkSonsLen(b, 2, c.config)
-    result = semBranchRange(c, t, b[0], b[1], covered)
-  else:
-    result = fitNode(c, t[0].typ, b, b.info)
-    inc(covered)
-
-proc semCaseBranch(c: PContext, t, branch: PNode, branchIndex: int,
-                   covered: var Int128) =
-  let lastIndex = branch.len - 2
-  for i in 0..lastIndex:
-    var b = branch[i]
+  result = semBranchRange(c, typ, b[1], b[2], covered)
+
+proc semCaseBranch(c: PContext, typ: PType, branch: PNode,
+                   covered: var Int128): PNode =
+  ## Analyses the ``nkOfBranch`` AST `branch`, producing either an error or
+  ## the typed AST. Note that the production does *not* include the action
+  ## slot.
+  ##
+  ## Productions with no items are valid: they result from, e.g., ``of {}:``.
+  proc fitAndAdd(to: PNode, c: PContext, typ: PType, n: PNode,
+                 covered: var Int128) {.nimcall.} =
+    if n.kind == nkError:
+      to.add n
+      return
+
+    # errors are ignored here; they're handled later
+    for it in n.items:
+      case it.kind
+      of nkRange:
+        to.add semBranchRange(c, typ, it[0], it[1], covered)
+      else:
+        var x = fitNode(c, typ, it, it.info)
+        # fitting the element may introduce a conversion
+        if x.kind in {nkHiddenCallConv, nkHiddenStdConv, nkHiddenSubConv}:
+          x = evalConstExpr(c, x)
+        to.add x
+        inc covered
+
+  checkMinSonsLen(branch, 2, c.config)
+  # not a one-to-one mapping between the input and output AST
+  result = copyNode(branch)
+
+  for i in 0..<branch.len-1:
+    let b = branch[i]
```

Comment:
```suggestion
  for i, b in branch.pairs:
```

That should be the same.

Reply:
Not quite, since the last slot needs to be skipped. I've now replaced the manual arithmetic with usage of the more robust `branchLabels` iterator.

Reply:
Ah, I totally zoned on the `<` and the `-1`, good call.